proposals/ref-readonly-parameters.md

```diff
@@ -0,0 +1,112 @@
+# `ref readonly` parameters
+
+## Summary
+[summary]: #summary
+
+Allow parameter declaration-site modifier `ref readonly` and change callsite rules as follows:
+
+| Callsite annotation  | `ref` parameter | `ref readonly` parameter | `in` parameter | `out` parameter |
+|----------------------|-----------------|--------------------------|----------------|-----------------|
+| `ref`                | Allowed         | **Allowed**              | **Warning**    | Error           |
+| `in`                 | Error           | **Allowed**              | Allowed        | Error           |
+| `out`                | Error           | **Error**                | Error          | Allowed         |
+| No annotation        | Error           | **Warning**              | Allowed        | Error           |
+
+Note that there is one change to the existing rules: `in` parameter with `ref` callsite annotation produces a warning instead of an error.
+
+| Value kind | `ref` parameter | `ref readonly` parameter | `in` parameter | `out` parameter |
+|------------|-----------------|--------------------------|----------------|-----------------|
+| rvalue     | Error           | **Warning**              | Allowed        | Error           |
+| lvalue     | Allowed         | **Allowed**              | Allowed        | Allowed         |
+
+## Motivation
+[motivation]: #motivation
+
+C# 7.2 [introduced `in` parameters](https://github.com/dotnet/csharplang/blob/c8c1615fcad4ca016a97b7260ad497aad53ebc78/proposals/csharp-7.2/readonly-ref.md#solution-in-parameters) as a way to pass readonly references.
+`in` parameters allow both lvalues and rvalues and can be used without any annotation at the callsite.
+However, APIs which capture or return references from their parameters would like to disallow rvalues and also enforce some indication at the callsite that a reference is being captured.
+`ref readonly` parameters are ideal in such cases as they warn if used with rvalues or without any annotation at the callsite.
+
+Furthermore, there are APIs that need only read-only references but use
+- `ref` parameters since they were introduced before `in` became available and changing to `in` would be a source and binary breaking change, e.g., `QueryInterface`, or
+- `in` parameters to accept readonly references even though passing rvalues to them doesn't really make sense, e.g., `ReadOnlySpan<T>..ctor(in T value)`, or
+- `ref` parameters to disallow rvalues even though they don't mutate the passed reference, e.g., `Unsafe.IsNullRef`.
+
+These APIs could migrate to `ref readonly` parameters without breaking users.
+For details on binary compatibility, see the proposed [metadata encoding](#metadata).
+Specifically, changing
+- `ref` → `ref readonly` would only be a binary breaking change for virtual methods,
```

Comment:
I do not think this is the right design choice. I think changing to `ref readonly` must be a binary breaking change regardless of virtual-ness. Languages that do not understand the semantics around `ref readonly` should be prevented form consuming APIs that use `ref readonly`. For example, it is common for VB to pass ByRef parameters via a temp with copy-back, which would violate the desired semantics around `ref readonly`.

Reply:
The same argument can be made for old versions of compilers (including C# compiler).

Reply:
The feature is effectively dead on arrival if it is a breaking change.

The BCL has numerous existing APIs that are desired to switch from `ref` to `ref readonly` and we cannot break the public surface area.

Reply:
> The feature is effectively dead on arrival if it is a breaking change.
 
I do not think so

>The BCL has numerous existing (non-virtual) APIs that are desired to switch from ref to ref readonly and we cannot break the existing public surface area.

The we shouldn't switch them.

Reply:
> Then we shouldn't switch them.

This feature is largely driven by a desire to do this switch. If the BCL can't make the switch then one of the biggest reasons to do the feature goes away and we would remain in a world where some of the most used APIs that would benefit from this can't use them, where users still have to manually workaround it using `ref T Unsafe.As(in T value)`.

Reply:
I believe it is a core requirement for the language to consider the 20+ years of existing code when introducing new features. If it doesn't then a large amount of value ends up being lost because the feature cannot correctly integrate and be used with the existing code that cannot make breaking changes due to back-compat.

Reply:
Personally, i'm struggling to see the value here if it woudl cause breaks.  It would likely need a host of new APIs introduced (and potential deprecation/hiding of the old APIs) and feels like it would boil-the-ocean in terms of getting people moved.  It feels more natural that this naturally work in a binary fashion, with warnings on recompile to get the people doing the rare-odd thing to move over.  If people miss the warnings, that's ok.  It's not the end of the world.  Their code was already busted before, so it's no worse.  But people that are paying attention can and who do care can gracefully move forward.

Reply:
> I think changing to ref readonly must be a binary breaking change regardless of virtual-ness.

That would defeat the purpose of this feature. The underlying motivation for this feature is allowing customers, particularly the dotnet/runtime team, the ability to migrate existing APIs which take `ref`, and to a lesser extent `in`, to a more correct API form of a non-mutating `ref`. If there is a binary compat break then that prevents existing APIs from moving over and hence it ceases it solve the underlying problem. Don't think there would be much appetite for that version of the feature.

> I believe there is a core need for the language to consider the 20+ years of existing code when introducing new features

Agree. At the time we designed `in` parameters we weren't as focused on adoption struggles as we should've been. It was also at the very start of the `Span<T>` and `readonly` movement so it was hard to forsee how much we'd move in that direction. If we were doing this feature today there are a number of tweaks in the design I believe we would consider.